### ISSUE TITLE
xetex issue: Improper discretionary list.

### ISSUE THREAD

**Issue by JulienPalard** [sphinx-doc/sphinx#5130]
Subject: Building pdf is failing with latex bug.

### Problem

I'm running sphinx-build on the french documentation translation of Python, and pdf build is failing.

#### Procedure to reproduce the problem
```
git clone https://github.com/python/python-docs-fr.git
git checkout 3.7
make MODE=autobuild-dev
```

#### Error logs / results
```
[...]
Writing index file howto-curses.idx

LaTeX Warning: Unused global option(s):
    [a4].

(./howto-curses.aux) (/usr/share/texlive/texmf-dist/tex/latex/base/ts1cmr.fd)
*geometry* driver: auto-detecting
*geometry* detected driver: xetex
(/usr/share/texlive/texmf-dist/tex/latex/hyperref/nameref.sty
(/usr/share/texlive/texmf-dist/tex/generic/oberdiek/gettitlestring.sty))

Package hyperref Warning: Rerun to get /PageLabels entry.

(/usr/share/texlive/texmf-dist/tex/latex/amsfonts/umsa.fd)
(/usr/share/texlive/texmf-dist/tex/latex/amsfonts/umsb.fd) [1] [2]
! Improper discretionary list.
<recently read> }
                 
l.264 ...gin\PYGZus{}y} \PYG{o}{=} \PYG{l+m+mi}{7}
                                                  
?  
! Emergency stop.
<recently read> }
                 
l.264 ...gin\PYGZus{}y} \PYG{o}{=} \PYG{l+m+mi}{7}
                                                  
Output written on howto-curses.pdf (2 pages).
Transcript written on howto-curses.log.
=== TeX engine is 'XeTeX'
Latexmk: Index file 'howto-curses.idx' was written
Latexmk: References changed.
Latexmk: Errors, so I did not complete making targets
Collected error summary (may duplicate other messages):
  pdflatex: Command for 'pdflatex' gave return code 1
      Refer to 'howto-curses.log' for details
Latexmk: Use the -f option to force complete processing,
 unless error was exceeding maximum runs, or warnings treated as errors.
Makefile:32 : la recette pour la cible « howto-curses.pdf » a échouée
make[3]: *** [howto-curses.pdf] Erreur 12
make[3] : on quitte le répertoire « /home/mdk/clones/cpython/Doc/build/latex »
Makefile:130 : la recette pour la cible « dist » a échouée
make[2]: *** [dist] Erreur 2
make[2] : on quitte le répertoire « /home/mdk/clones/cpython/Doc »
Makefile:187 : la recette pour la cible « autobuild-dev » a échouée
make[1]: *** [autobuild-dev] Erreur 2
make[1] : on quitte le répertoire « /home/mdk/clones/cpython/Doc »
Makefile:28: recipe for target 'all' failed
make: *** [all] Error 2
```

#### Expected results
I expected PDF builds to succeed :)

### Environment info
- OS: Debian buster
- Python version:3.7
- Sphinx version:1.7.5
- texlive-xetex: 2018.20180505-1

**Comment by jfbu** [comment]
I confirm this. Pending closer scrutiny this reminds me of #3546 which was closed as wont-fix because it revealed a XeTeX bug.

I have confirmed using lualatex on `using.tex` file has no issue. Currently trying building the full docs with `lualatex`.

**Comment by jfbu** [comment]
I could build all PDFs with `lualatex` but another issue arose.

The PDF builds was broken for tutorial.pdf, library.pdf, reference.pdf due to indexing of word ``état``. This is caused by `makeindex` not being utf-8 savvy.

A correct utf-8 file `tutorial.idx` is parsed by makeindex which creates a wrong `tutorial.ind` because it took the first byte of ``é``. Then this breaks with this kind of error

```
! String contains an invalid utf-8 sequence.
l.9745   \bigletter 
                  ?^^@
? 

```

(this was rather for library.tex I think).

I located in the tex files all ``\index{état!....}`` and variants and commented them out, then `make all-pdf` in `build/latex` succeeded. (after removing first all idx and ind files)

However obviously the better solution would be to configure Sphinx to use `xindy` as indexing engine not `makeindex`. This should not be too difficult because the ``latexmkrc``file can be overridden by Sphinx users and configured to use ``xindy``. [This link](https://tex.stackexchange.com/a/199841/4686) gives many gory LaTeX details and makindex vs xindy. At time of writing I am confused how to let ``xindy`` use the ``python.ist`` style file of makeindex, or if I need to provide another one (apparently with file suffix ``.xdy`` not ``.ist``).

I will need a little time to clarify how to use `xindy` with Sphinx and UTF-8 engines. Perhaps I can do this in a few hours. If not PRs are always welcome. (I have deadlines)

I will investigate the XeTeX issue later, because there is nothing I can do to fix that issue except report it to XeTeX maintainers.

**Comment by jfbu** [comment]
For the record, I have digged out this [report](http://tug.org/pipermail/xetex/2017-December/027212.html) where I tried to get some information on #3546. This reminded me there was a problem with xelatex+polyglossia+french, which was mentioned in this (technical) TeX thread.

Sure enough if I use in `latex_elements`, `'babel': '\\usepackage{babel}'`, then I can compile with latex_engine set to xelatex with no problem this reST source:

```
When invoking Python, you may specify any of these options::

    python [-bBdEhiIOqsSuvVWx?] [-c command | -m module-name | script | - ] [args]

```

I extracted this source as minimal example causing the error you reported for PDF builds if

- using xelatex,

- AND language is French.

So you have probably hit against some polyglossia+french bug, which we can't help with at Sphinx. It might be the bug https://github.com/reutenauer/polyglossia/issues/145 which is supposedly fixed by already merged https://github.com/reutenauer/polyglossia/pull/178 but a release of polyglossia is long awaited. (I have not tested this does fix our issue with above source). At least that bug is the one mentioned in the [TeX thread](http://tug.org/pipermail/xetex/2017-December/027212.html) I initiated a few months back (and got no answer to my initial query).

I propose two workarounds:

- use  the ``'babel'`` key like this:

```
latex_elements = {
    # The paper size ('letterpaper' or 'a4paper').
    #
    # 'papersize': 'letterpaper',

    # The font size ('10pt', '11pt' or '12pt').
    #
    'babel': '\\usepackage{babel}',

    # Additional stuff for the LaTeX preamble.
    #
    # 'preamble': '',

    # Latex figure (float) alignment
    #
    # 'figure_align': 'htbp',
}
```

- or use latex_engine = 'lualatex'.

But both solutions (I tested only the lualatex as reported above) will then give rise to the second problem, the one with the indexing of `état`.

**Comment by jfbu** [comment]
ah wait, the problem with indexing of `état` is silently ignored by `xelatex` (contrarily to `lualatex`). (the `.ind` files are not valid utf-8 files, but `xelatex` creates the index with no complaint, as this shows:

![capture d ecran 2018-06-29 a 23 44 46](https://user-images.githubusercontent.com/2589111/42116189-79110442-7bf6-11e8-8fa3-7b08eaafd008.png)

There should be a capital `É` as heading, but the file only contains the first byte of its utf-8 encoding... nevertheless xelatex does not stop compilation. In the same circumstance lualatex throws an error.

So you can fix your problem by using `\usepackage{babel}` in place of polyglossia.

@tk0miya  this relates my #4272. It would make sense to swicth for xelatex/lualatex from polyglossia to babel in case of French language.

**Comment by jfbu** [comment]
For the record, I mention that ``.. index::  état`` does not created failed PDF builds with standard pdflatex engine. In that case the ``é`` is converted into macros by LaTeX, and makeindex works with something like `\IeC {\'e}` which it sorts according to first character, backslash, so there is no problem of file encoding.

**Comment by jfbu** [comment]
@JulienPalard I am closing this because it is fixed by #5178 which is merged into soon to be released 1.7.6.

In building CPython docs and examining indices I also met issue #5176 which showed up in indexing of ``!=`` operator for example. The #5177 fix is also merged into future release 1.7.6.

However Sphinx uses `makeindex` for LaTeX general index, and words like ``état`` are not located at nice place in index. In future this will be fixed by #5134 which will be as new feature released with 1.8: Sphinx will then use by default `xindy` for `xelatex`.
